### PR TITLE
Write out cross-validation results immediately

### LINKED ISSUE #2480: Write cross-validation results immediately after cross validation
**Please describe the use case that requires this feature.**
Currently, the `CrossValidator` keeps the results of the cross-validation, including the full result table in memory and writes it in `finish` method of the tools.

We should write the results immediately in the cross validation loop using `write_table(..., append=True)` to avoid keeping this information in memory unnecessarily.

**Describe the solution you'd like**

Write cross validation results out immediately after each step, not in `finish`.

### PR DESCRIPTION
**Pull request by LukasBeiske**
Fixes #2480

### CODE REVIEW

File: ctapipe/reco/sklearn.py [hunk 1]
```diff
@@ -878,9 +878,8 @@ class CrossValidator(Component):
         default_value=1337, help="Random seed for splitting the training data."
     ).tag(config=True)
 
-    def __init__(self, model_component, **kwargs):
+    def __init__(self, model_component, overwrite=False, **kwargs):
```
Comment (maxnoe) [inline ctapipe/reco/sklearn.py:881]: You don't use overwrite

File: ctapipe/reco/sklearn.py [hunk 2]
```diff
@@ -965,7 +967,14 @@ def __call__(self, telescope_type, table):
...
                     cv_values.std(),
                 )
-        self.cv_predictions[telescope_type] = vstack(predictions)
+
+        if self.output_path:
+            write_table(
+                vstack(predictions),
+                self.output_path,
+                f"/cv_predictions_{telescope_type}",
+                overwrite=True,
```
Comment (maxnoe) [inline ctapipe/reco/sklearn.py:976]: And here you set it to true immediately

File: ctapipe/reco/sklearn.py (hunk 1 above)
Comment (LukasBeiske) [reply ctapipe/reco/sklearn.py:881]: I forgot to remove that when removing the redundant test 👍

File: ctapipe/reco/sklearn.py (hunk 2 above)
Comment (maxnoe) [reply ctapipe/reco/sklearn.py:976]: You can move the writing into the loop (further reducing memory) by adding `append=True` here after overwriting the file in `__init__`

**Review by maxnoe — CHANGES_REQUESTED** [review]
Good start, but you should

a) Check overwrite conditions in `__init__`: if output file exists and not overwrite, raise an error, if output file exists and overwrite=true truncate it (tables.open_file(path, mode='w'). 

b) Move the writing into the cv loop using `write_table(..., append=True)` to further reduce memory usage

c) Add writing of metadata to the output file

**Comment by LukasBeiske** [comment]
> Good start, but you should
> 
> a) Check overwrite conditions in `__init__`: if output file exists and not overwrite, raise an error, if output file exists and overwrite=true truncate it (tables.open_file(path, mode='w').
> 
> b) Move the writing into the cv loop using `write_table(..., append=True)` to further reduce memory usage
> 
> c) Add writing of metadata to the output file

a) The error if file exists and not overwrite is already raised by the train tools via `Tool.check_output`.

a)/b) Using `write_table(..., append=True, mode='w')` does not work, does it? Isn't this basically the same as `write_table(..., append=True, overwrite=True)`?

c) Is there any other metadata requiered besides some column descriptions?

**Comment by maxnoe** [comment]
> a) The error if file exists and not overwrite is already raised by the train tools via Tool.check_output.

Also for the cv output file and not just the model?

**Comment by maxnoe** [comment]
> a)/b) Using write_table(..., append=True, mode='w') does not work, does it? Isn't this basically the same as write_table(..., append=True, overwrite=True)?

You need to truncate the file once in `__init__` and then only use `append=True`

**Comment by LukasBeiske** [comment]
> > a) The error if file exists and not overwrite is already raised by the train tools via Tool.check_output.
> 
> Also for the cv output file and not just the model?

Yes, both output paths are passed and `CrossValidator.ouput_path` gets ignored if it is None

File: ctapipe/reco/sklearn.py [hunk 3]
```diff
@@ -898,6 +905,29 @@ def __init__(self, model_component, **kwargs):
                 "Unsupported Model of type %s supplied", self.model_component
             )
 
+        if self.output_path:
```
Comment (Tobychev) [inline ctapipe/reco/sklearn.py:908]: You can combine these two into `if self.output_path and self.output_path.exists():` without any danger, if `output_path` is none the second bit is not executed.

Comment (LukasBeiske) [reply ctapipe/reco/sklearn.py:908]: But if I do it like this, the lines below the checks for the correct configuration regarding `overwrite` do not get executed if the output path doesn't already exist. So I would have to put those in a separate if statement which would result in the same number of if statements as right now.

File: ctapipe/reco/sklearn.py [hunk 4]
```diff
@@ -937,21 +965,34 @@ def __call__(self, telescope_type, table):
...
-                    {
-                        "cv_fold": np.full(len(truth), fold, dtype=np.uint8),
-                        "tel_type": [str(telescope_type)] * len(truth),
-                        "truth": truth,
-                        "true_energy": test["true_energy"],
-                        "true_impact_distance": test["true_impact_distance"],
-                        **cv_prediction,
-                    }
+            cv_result, metrics = self.cross_validate(telescope_type, train, test)
+            if self.output_path:
```
Comment (maxnoe) [inline ctapipe/reco/sklearn.py:969]: This is very unreadable... introducing some intermediate variables might be better, i.e:
```
results = Table(
     ...
)
results = hstack([results, cv_predictions], joint_type='exact')
```

**Review by Tobychev — APPROVED** [review]

**Pull request merged**